src/keyring.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::HashMap;
+
+use async_std::sync::Mutex;
+use zeroize::Zeroizing;
+
+use crate::{
+    dbus::{self, Algorithm, DEFAULT_COLLECTION},
+    portal, Result,
+};
+
+/// A generic implementation of Keyring.
+///
+/// It would automatically use the file backend if the application is sandboxed
+/// and otherwise falls back to the DBus service.
+pub struct Keyring {
```

Comment:
Why not enum?

Reply:
Because it contains the is_sandboxed field but sure, an enum could work indeed